src/swap/db.rs

```diff
@@ -9,3 +16,757 @@ pub trait Load<T>: Send + Sync + 'static {
+    }
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+struct Swap {
+    pub hbit_funded: Option<HbitFunded>,
+    pub hbit_redeemed: Option<HbitRedeemed>,
+    pub hbit_refunded: Option<HbitRefunded>,
+    pub herc20_deployed: Option<Herc20Deployed>,
+    pub herc20_funded: Option<Herc20Funded>,
+    pub herc20_redeemed: Option<Herc20Redeemed>,
+    pub herc20_refunded: Option<Herc20Refunded>,
+}
+
+impl Default for Swap {
+    fn default() -> Self {
+        Swap {
+            hbit_funded: None,
+            hbit_redeemed: None,
+            hbit_refunded: None,
+            herc20_deployed: None,
+            herc20_funded: None,
+            herc20_redeemed: None,
+            herc20_refunded: None,
+        }
+    }
+}
+
+// TODO: control the serialisation
+#[derive(Clone, Debug, Serialize, Deserialize)]
+struct HbitFunded {
+    pub asset: u64,
+    pub location: ::bitcoin::OutPoint,
+}
+
+impl From<HbitFunded> for hbit::Funded {
+    fn from(funded: HbitFunded) -> Self {
+        hbit::Funded {
+            asset: comit::asset::Bitcoin::from_sat(funded.asset),
+            location: funded.location,
+        }
+    }
+}
+
+impl From<hbit::Funded> for HbitFunded {
+    fn from(funded: hbit::Funded) -> Self {
+        HbitFunded {
+            asset: funded.asset.as_sat(),
+            location: funded.location,
+        }
+    }
+}
+
+// Kind of bending the arm of the trait
+#[async_trait::async_trait]
+impl Save<Created> for Database {
+    async fn save(&self, _event: Created, swap_id: SwapId) -> anyhow::Result<()> {
+        let stored_swap = self.get(&swap_id);
+
+        match stored_swap {
+            Ok(_) => Err(anyhow!("Swap is already stored")),
+            Err(_) => {
+                let swap = Swap::default();
+                let new_value =
+                    serde_json::to_vec(&swap).context("Could not serialize new swap value")?;
+
+                self.db
+                    .compare_and_swap(swap_id.as_bytes(), Option::<Vec<u8>>::None, Some(new_value))
+                    .context("Could not write in the DB")?
+                    .context("Stored swap somehow changed, aborting saving")
+            }
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Save<hbit::Funded> for Database {
+    async fn save(&self, event: hbit::Funded, swap_id: SwapId) -> anyhow::Result<()> {
+        let stored_swap = self.get(&swap_id)?;
+
+        match stored_swap.hbit_funded {
+            Some(_) => Err(anyhow!("Hbit Funded event is already stored")),
+            None => {
+                let mut swap = stored_swap.clone();
+                swap.hbit_funded = Some(event.into());
+
+                let old_value = serde_json::to_vec(&stored_swap)
+                    .context("Could not serialize old swap value")?;
+                let new_value =
+                    serde_json::to_vec(&swap).context("Could not serialize new swap value")?;
+
+                self.db
```

Comment:
:question: Why is `compare_and_swap` preferred over plain old `insert`?

Reply:
Just using the recommended API of sled :)